library/src/main/java/io/realm/RealmBasedRecyclerViewAdapter.java

```diff
@@ -353,8 +359,22 @@ public void updateRealmResults(RealmResults<T> queryResults) {
      * Method that creates the header string that should be used. Override this method to have
      * a custom header.
      */
-    public String createHeaderFromColumnValue(String columnValue) {
-        return columnValue.substring(0, 1);
+    public String createHeaderFromColumnValue(Object columnValue) {
```

Comment:
Could you change this to throw an exception if the `columnValue` is not supported? 

Such as

  if (columnValue instanceof Boolean) {
    return columnValue.toString();
  } else if (columnValue instanceof String) {
  ....
  } else if (...) {
   ....
  }
  throw new IllegalStateException("columnValue type not supported");


This should make it easy to catch / more explicit than returning null.